REGRESSION (287736@main?): [ iOS ] fast/mediastream/microphone-interruption-and-audio-session.html is constantly failing.

youennf opened the pull request:
#### 755ea5bdabcb1ab6b7e6135f7de2370f9cbe4909
<pre>
REGRESSION (287736@main?): [ iOS ] fast/mediastream/microphone-interruption-and-audio-session.html is constantly failing.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287033">https://bugs.webkit.org/show_bug.cgi?id=287033</a>
<a href="https://rdar.apple.com/144177133">rdar://144177133</a>

Reviewed by Jean-Yves Avenard.

The test is flaky as we sometimes recreate the GPUProcess and the CoreAudioCaptureSourceFactory will register itself as an interruption observer to the dummy AudioSession.
Instead, we are now having a singleton set of interruption observers, used by AudioSession objects when being interrupted/ending interruptions.
Only the shared AudioSession is expected to being interrupted/ending interruptions, so there should be no issue with several AudioSessions doing interruptions in parallel.
We also update the test to make sure capture tracks are stopped explicitly.

* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::DOMAudioSession):
(WebCore::DOMAudioSession::~DOMAudioSession):
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::audioSessionInterruptionObserversSingleton):
(WebCore::AudioSession::addInterruptionObserver):
(WebCore::AudioSession::removeInterruptionObserver):
(WebCore::AudioSession::beginInterruption):
(WebCore::AudioSession::endInterruption):
(WebCore::AudioSession::activeStateChanged):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
(WebCore::MediaSessionManageriOS::MediaSessionManageriOS):
(WebCore::MediaSessionManageriOS::~MediaSessionManageriOS):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSourceFactory::CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::~CoreAudioCaptureSourceFactory):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::RemoteAudioSessionProxyManager):
(WebKit::RemoteAudioSessionProxyManager::~RemoteAudioSessionProxyManager):
(WebKit::RemoteAudioSessionProxyManager::beginInterruptionRemote):
(WebKit::RemoteAudioSessionProxyManager::endInterruptionRemote):
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::~RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::RemoteAudioSession):
(WebKit::RemoteAudioSession::~RemoteAudioSession):
(WebKit::RemoteAudioSession::beginInterruptionRemote):
(WebKit::RemoteAudioSession::endInterruptionRemote):

Canonical link: <a href="https://commits.webkit.org/300015@main">https://commits.webkit.org/300015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/469ad34e69bf113a213c0103d55c325c07f4ed2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40182 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30834 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126864 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72563 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a3f73229-d2c6-4244-864a-08e60657adfb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122364 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48759 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91510 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60790 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2a9580c9-d2e9-41a6-bdc1-fe204cc5188f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32649 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108009 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72061 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8c4bde97-b2cb-4d43-b7e2-03f5c5064202) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31682 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26115 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70481 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102133 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26297 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129749 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47409 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35990 "Found 1 new test failure: http/tests/websocket/tests/hybi/alert-in-event-handler.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100129 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47776 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99971 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25465 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45418 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23446 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44057 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47271 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52976 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46739 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50086 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48426 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/audio/AudioSession.cpp:
```diff
@@ -152,14 +152,20 @@ bool AudioSession::tryToSetActive(bool active)
 }
 
+static WeakHashSet<AudioSessionInterruptionObserver>& audioSessionInterruptionObserversSingleton()
+{
```
can we have an `assertIsMainThread()` here ?

jyavenard commented on Source/WebCore/platform/audio/AudioSession.cpp:
```diff
@@ -152,14 +152,20 @@ bool AudioSession::tryToSetActive(bool active)
 void AudioSession::addInterruptionObserver(AudioSessionInterruptionObserver& observer)
 {
-    m_interruptionObservers.add(observer);
+    audioSessionInterruptionObserversSingleton().add(observer);
```
that's a mouthful !

youennf committed:
REGRESSION (287736@main?): [ iOS ] fast/mediastream/microphone-interruption-and-audio-session.html is constantly failing.
https://bugs.webkit.org/show_bug.cgi?id=287033
rdar://144177133

Reviewed by Jean-Yves Avenard.

The test is flaky as we sometimes recreate the GPUProcess and the CoreAudioCaptureSourceFactory will register itself as an interruption observer to the dummy AudioSession.
Instead, we are now having a singleton set of interruption observers, used by AudioSession objects when being interrupted/ending interruptions.
Only the shared AudioSession is expected to being interrupted/ending interruptions, so there should be no issue with several AudioSessions doing interruptions in parallel.
We also update the test to make sure capture tracks are stopped explicitly.

* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::DOMAudioSession):
(WebCore::DOMAudioSession::~DOMAudioSession):
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::audioSessionInterruptionObserversSingleton):
(WebCore::AudioSession::addInterruptionObserver):
(WebCore::AudioSession::removeInterruptionObserver):
(WebCore::AudioSession::beginInterruption):
(WebCore::AudioSession::endInterruption):
(WebCore::AudioSession::activeStateChanged):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
(WebCore::MediaSessionManageriOS::MediaSessionManageriOS):
(WebCore::MediaSessionManageriOS::~MediaSessionManageriOS):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSourceFactory::CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::~CoreAudioCaptureSourceFactory):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::RemoteAudioSessionProxyManager):
(WebKit::RemoteAudioSessionProxyManager::~RemoteAudioSessionProxyManager):
(WebKit::RemoteAudioSessionProxyManager::beginInterruptionRemote):
(WebKit::RemoteAudioSessionProxyManager::endInterruptionRemote):
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::~RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::RemoteAudioSession):
(WebKit::RemoteAudioSession::~RemoteAudioSession):
(WebKit::RemoteAudioSession::beginInterruptionRemote):
(WebKit::RemoteAudioSession::endInterruptionRemote):

Canonical link: https://commits.webkit.org/300015@main
